[Lens][Agg based table]Navigate to lens for agg based table vis

VladLasitsa opened the pull request:
## Summary
Completes part of https://github.com/elastic/kibana/issues/138236.

As part of phasing out TSVB and Visualize all Legacy agg based visulizations should support "open in lens" functionality.
In that PR converter for Agg based table was added.

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

spalger reviewed (approved):
Lens bundle bump LGTM

stratoula commented:
@VladLasitsa I see +18KB to visualizations bundle. Can we fix this?

Kunzetsov commented:
> @VladLasitsa I see +18KB to visualizations bundle. Can we fix this?

Hey, @stratoula, I'll try to fix this. I guess I've added some extra exports.

Kunzetsov commented:
@elastic/kibana-app-services and @elastic/kibana-vis-editors, could you, please, review this PR? Thanks a lot.

mbondyra reviewed (commented):
Started testing, will keep add potential problems in this comment 

1. Only one column count table causes an error: 
<img width="914" alt="Screenshot 2022-09-21 at 14 29 41" src="https://user-images.githubusercontent.com/4283304/191504045-efb20e68-353a-4ee1-9a88-5e90dcd25e3c.png">
<img width="914" alt="Screenshot 2022-09-21 at 14 29 46" src="https://user-images.githubusercontent.com/4283304/191504057-21dbe3ce-34f2-4de8-bae7-c50c7f822ba1.png">

Kunzetsov commented:
@mbondyra, I've changed the way we are receiving indexPatternIds from `navigateToLens` context. Could you, please, review it when you'll have some time? Thanks a lot.

ppisljar commented on src/plugins/data/common/search/aggs/types.ts:
```diff
@@ -156,6 +169,50 @@ export type AggExpressionFunctionArgs<Name extends keyof AggParamsMapping> =
  *
  * @internal
  */
+export interface SerializedAggParamsMapping {
```
i can't find this used anywhere but in this same file above ?
- do we need to export it ?
- whats the intention of this change on agg configs ?

Kunzetsov replied:
- You are right, it makes sense not to export this interface.
- `SerializedAggParamsMapping` is representing args, which are passed to the expression functions, but `SchemaConfig` contains params of different types.
Good example: 
https://github.com/elastic/kibana/blob/3494b2af1e961ddcc7233bdd6bb184751325af4d/src/plugins/data/common/search/aggs/metrics/bucket_avg.ts#L18
https://github.com/elastic/kibana/blob/3494b2af1e961ddcc7233bdd6bb184751325af4d/src/plugins/data/common/search/aggs/metrics/bucket_avg.ts#L23

And later usage: 
<img width="499" alt="Screenshot 2022-09-22 at 10 35 43" src="https://user-images.githubusercontent.com/22456368/191686358-b85d5062-6ba5-4c14-9341-8b83027b9b5d.png">
<img width="499" alt="Screenshot 2022-09-22 at 10 35 56" src="https://user-images.githubusercontent.com/22456368/191686390-597b8bc8-56d5-4710-8f8d-209a0d52cb65.png">

We had to add types to the aggParams to make the code more predictable and documented.

mbondyra reviewed (commented):
1. The operations that allow dates (or geofield) in datatable but not in Lens cause improper dimensions created. I think we should either remove them or not allow to convert to lens when they occur. Here’s the 3 examples:

<img width="1029" alt="Screenshot 2022-09-22 at 11 13 13" src="https://user-images.githubusercontent.com/4283304/191708492-88b0904f-d26c-4181-807e-c31e53a60022.png">

2. Also, a question - why do we block histogram and ranges from converting? 🤔

mbondyra commented on src/plugins/vis_types/table/public/convert_to_lens/configurations/index.ts:
```diff
@@ -0,0 +1,67 @@
+
+const getPagination = ({ perPage }: TableVisParams): PagingState => {
+  return {
+    enabled: perPage !== '' ? true : false,
```
```suggestion
    enabled: perPage !== '' ,
```

mbondyra commented:
One more that we might want to consider to convert is:

<img width="1114" alt="Screenshot 2022-09-22 at 13 47 22" src="https://user-images.githubusercontent.com/4283304/191739862-73468e6a-8560-44ce-93ff-a0a41735aac7.png">

to:

<img width="907" alt="Screenshot 2022-09-22 at 13 53 13" src="https://user-images.githubusercontent.com/4283304/191739921-2ec0a7a0-4a17-4c83-b18c-3a548bfe3f12.png">

mbondyra commented on src/plugins/visualizations/common/convert_to_lens/lib/convert/last_value.ts:
```diff
@@ -0,0 +1,61 @@
+): LastValueParams => {
+  return {
+    sortField: agg.aggParams!.sortField!.name,
+    showArrayValues: agg.aggType === METRIC_TYPES.TOP_HITS ? true : false,
```
```suggestion
    showArrayValues: agg.aggType === METRIC_TYPES.TOP_HITS,
```

VladLasitsa commented:
> One more that we might want to consider to convert is:
> 
> <img alt="Screenshot 2022-09-22 at 13 47 22" width="1114" src="https://user-images.githubusercontent.com/4283304/191739862-73468e6a-8560-44ce-93ff-a0a41735aac7.png">
> 
> to:
> 
> <img alt="Screenshot 2022-09-22 at 13 53 13" width="907" src="https://user-images.githubusercontent.com/4283304/191739921-2ec0a7a0-4a17-4c83-b18c-3a548bfe3f12.png">

@flash1293, Could you please comment this? in document we have this text: "_block conversion if there are multiple average/min/max/sum buckets with different bucket aggs or there is a split series defined_". Because of that we block conversion when sibling pipeline agg and split row defined.

flash1293 commented:
> in document we have this text: "block conversion if there are multiple average/min/max/sum buckets with different bucket aggs or there is a split series defined". Because of that we block conversion when sibling pipeline agg and split row defined.

@VladLasitsa I think I worded this badly - the case @mbondyra mentions should be converted. What I meant is something like this:
<img width="2540" alt="Screenshot 2022-09-22 at 14 25 04" src="https://user-images.githubusercontent.com/1508364/191746076-be27b4b3-686f-4933-911c-d0d15e2e3e81.png">

Using an average bucket and a sum bucket in a single chart. This can't be expressed in Lens with the "collapse by" row dimensions as they are applied to all metrics in the same way.

VladLasitsa commented:
> > in document we have this text: "block conversion if there are multiple average/min/max/sum buckets with different bucket aggs or there is a split series defined". Because of that we block conversion when sibling pipeline agg and split row defined.
> 
> @VladLasitsa I think I worded this badly - the case @mbondyra mentions should be converted. What I meant is something like this: <img alt="Screenshot 2022-09-22 at 14 25 04" width="2000" src="https://user-images.githubusercontent.com/1508364/191746076-be27b4b3-686f-4933-911c-d0d15e2e3e81.png">
> 
> Using an average bucket and a sum bucket in a single chart. This can't be expressed in Lens with the "collapse by" row dimensions as they are applied to all metrics in the same way.

Got it, Added support of this case.

Kunzetsov commented:
@elasticmachine merge upstream

spalger commented:
@elasticmachine merge upstream

(to resolve ci issues caused by Github outage)

mbondyra reviewed (approved):
There are two small inconsistencies that can be target later if at all:
1. The histogram is not giving the same intervals as converted intervals:
<img width="488" alt="Screenshot 2022-09-23 at 12 09 20" src="https://user-images.githubusercontent.com/4283304/191938859-308a29c4-64d9-4364-87e8-018cee1ab2c0.png">
<img width="489" alt="Screenshot 2022-09-23 at 12 09 25" src="https://user-images.githubusercontent.com/4283304/191938823-1f47f355-c73a-44fd-aa9b-e519e29098e6.png">

2. For the field types, it doesn't allow to convert for simple metric if used the not suppored fieldtype, like this one: 
<img width="474" alt="Screenshot 2022-09-23 at 12 00 58" src="https://user-images.githubusercontent.com/4283304/191938975-16a9e6f4-4b67-41eb-aac1-3bf189e1eaa9.png">

But for cumulative_sum of the same metric it allows to convert and then shows nice readable error message:
<img width="491" alt="Screenshot 2022-09-23 at 12 01 15" src="https://user-images.githubusercontent.com/4283304/191939110-66715097-6a91-49e5-8203-f65fa8a68f45.png">
converts to:
<img width="331" alt="Screenshot 2022-09-23 at 12 01 19" src="https://user-images.githubusercontent.com/4283304/191939146-e243aa18-e898-490b-8773-8db2588eb991.png">

I find neither of them critical so approving. It can be addressed later.

VladLasitsa commented:
> There are two small inconsistencies that can be target later if at all:
> 
> 1. The histogram is not giving the same intervals as converted intervals:
> 
> <img alt="Screenshot 2022-09-23 at 12 09 20" width="488" src="https://user-images.githubusercontent.com/4283304/191938859-308a29c4-64d9-4364-87e8-018cee1ab2c0.png">
> 
> <img alt="Screenshot 2022-09-23 at 12 09 25" width="489" src="https://user-images.githubusercontent.com/4283304/191938823-1f47f355-c73a-44fd-aa9b-e519e29098e6.png">
> 

@mbondyra  About that: we don't have possibility to provide interval for "intervals" in Lens.
We have only these params:
<img width="644" alt="Снимок экрана 2022-09-23 в 14 08 51" src="https://user-images.githubusercontent.com/16915480/191947969-cda46f73-1451-4aa4-9958-29dae012ad3c.png">
In Lens implementation we always provide "auto" as interval for histogram:

<img width="646" alt="Снимок экрана 2022-09-23 в 14 10 41" src="https://user-images.githubusercontent.com/16915480/191948195-afb79824-8d63-4120-931c-6437d117402e.png">

After discussion with Joe we decided to ignore interval for histogram.

cc @flash1293

Kunzetsov commented:
> 2. For the field types, it doesn't allow to convert for simple metric if used the not suppored fieldtype, like this one:
> 
> <img alt="Screenshot 2022-09-23 at 12 00 58" width="474" src="https://user-images.githubusercontent.com/4283304/191938975-16a9e6f4-4b67-41eb-aac1-3bf189e1eaa9.png">
> 
> But for cumulative_sum of the same metric it allows to convert and then shows nice readable error message: <img alt="Screenshot 2022-09-23 at 12 01 15" width="491" src="https://user-images.githubusercontent.com/4283304/191939110-66715097-6a91-49e5-8203-f65fa8a68f45.png"> converts to: <img alt="Screenshot 2022-09-23 at 12 01 19" width="331" src="https://user-images.githubusercontent.com/4283304/191939146-e243aa18-e898-490b-8773-8db2588eb991.png">

@mbondyra, we've fixed this problem.

mbondyra commented:
Rechecked, and re-approving!

stratoula reviewed (changes requested):
Sorry for being so annoying with the sizes but + 24KB are a lot :) Can we look into it?

<img width="500" alt="image" src="https://user-images.githubusercontent.com/17003240/191987675-d5f397a6-39c0-4fb2-bafe-41a4d6272b35.png">

ppisljar reviewed (approved):
LGTM

Kunzetsov commented:
@elasticmachine merge upstream
